Feature/validation test preformance

saied2035 opened the pull request:
### Project requirements
- I added the following validations:
  - For the User model:
    - Name must not be blank.
    - PostsCounter must be an integer greater than or equal to zero.
  - For the Post model:
    - Title must not be blank.
    - Title must not exceed 250 characters.
    - CommentsCounter must be an integer greater than or equal to zero.
    - LikesCounter must be an integer greater than or equal to zero.
- I added unit specs for all of the models' methods and validations.
- I added flash messages in the create actions in all the controllers.
- The N+1 problem is solved when fetching all posts and their comments for a user.

Thanks for reviewing my PR 😄

saied2035 committed:
Merge branch 'feature/add-forms' into dev

saied2035 committed:
use includes for better preformance

saied2035 committed:
add validations

saied2035 committed:
add error flash messages for all create methods in controllers

saied2035 committed:
finish test for comment model

saied2035 committed:
finish test for like model

saied2035 committed:
finish test for post model

saied2035 committed:
finish test for user model

saied2035 committed:
fix linter errors

Lameck1 reviewed (changes requested):
### Hi, @saied2035, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights
- Professional README ✔️
- Clean code ✔️ 
- Good commit messages ✔️ 

## Required Changes ♻️
- [x] I could not find `posts_controller.rb` in the `Files Changed` tab, so I am gonna leave this here 👇🏼:

  One of the project requirements states this:
  > - Make sure that the N+1 problem is solved when fetching all posts and their comments for a user.
  
  The above-quoted requirement doesn't seem to have been taken care of. This is what you have so far in the `index` action of `posts_controller`:
  
  ```ruby
  def index
    @user = all_posts
    @posts = @user.posts.order('id asc')
  end
  ```
  We need to fetch comments too, alongside the posts. Consider this 👇🏼:
  
  ```ruby
  @posts = @user.posts.includes(:comments)
  ```
  Please work on this.
  
  ## A word of concern
  - Consider giving you methods/functions names that tell what they actually do. For instance, you named a method that finds and returns the user `all_posts`. I was then wondering, how is `@user = all_posts`? A method named `all_posts` should logically return all posts as the name suggests.
  
  - Giving methods proper is necessary for code readability and maintainability.

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
**Please, remember to tag me in your question so I can receive the notification.**

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

saied2035 committed:
finish required changes

Lameck1 reviewed (approved):
### Hi, @saied2035,

You've done well addressing concerns from the previous review 👏🏼.
Tasks for this project are now successfully implemented 💯.

## Highlights
- Professional README ✔️
- Clean code ✔️ 
- Good commit messages ✔️ 

All the best with the next project.

You are free to merge this branch :shipit:
Congratulations! 🎉 

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
**Please, remember to tag me in your question so I can receive the notification.**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

saied2035 commented:
@Lameck1  I appreciate your review a lot. I didn't notice the `N+1` solution should be placed in the `post controller` so I moved the solution to the `post controller`—also, thanks for notifying me about the method name. I forgot to change it before I submit. I changed it to a better name. Thank again for your honest review. 😄
